aggregator/election_mgr.go

```diff
@@ -0,0 +1,505 @@
+		resignNotFollower:    scope.Counter("resign-not-follower"),
+		followerToLeader:     scope.Counter("follower-to-leader"),
+		leaderToFollower:     scope.Counter("leader-to-follower"),
+		electionState:        scope.Gauge("election-state"),
+	}
+}
+
+type electionManager struct {
+	sync.RWMutex
+
+	nowFn           clock.NowFn
+	logger          xlog.Logger
+	reportInterval  time.Duration
+	campaignOpts    services.CampaignOptions
+	electionOpts    services.ElectionOptions
+	campaignRetrier xretry.Retrier
+	changeRetrier   xretry.Retrier
+	electionKeyFmt  string
+	leaderService   services.LeaderService
+	leaderValue     string
+
+	state            electionManagerState
+	doneCh           chan struct{}
+	resignWg         *sync.WaitGroup
+	electionKey      string
+	electionState    ElectionState
+	changeInProgress bool
+	changeCancelFn   context.CancelFunc
+	changeWg         sync.WaitGroup
+	sleepFn          sleepFn
+	metrics          electionManagerMetrics
+}
+
+// NewElectionManager creates a new election manager.
+func NewElectionManager(opts ElectionManagerOptions) ElectionManager {
+	instrumentOpts := opts.InstrumentOptions()
+	campaignOpts := opts.CampaignOptions()
+	campaignRetrier := xretry.NewRetrier(opts.CampaignRetryOptions().SetForever(true))
+	changeRetrier := xretry.NewRetrier(opts.ChangeRetryOptions().SetForever(true))
+	mgr := &electionManager{
+		nowFn:           opts.ClockOptions().NowFn(),
+		logger:          instrumentOpts.Logger(),
+		reportInterval:  instrumentOpts.ReportInterval(),
+		campaignOpts:    campaignOpts,
+		electionOpts:    opts.ElectionOptions(),
+		campaignRetrier: campaignRetrier,
+		changeRetrier:   changeRetrier,
+		electionKeyFmt:  opts.ElectionKeyFmt(),
+		leaderService:   opts.LeaderService(),
+		leaderValue:     campaignOpts.LeaderValue(),
+		state:           electionManagerNotOpen,
+		doneCh:          make(chan struct{}),
+		electionState:   FollowerState,
+		sleepFn:         time.Sleep,
+		metrics:         newElectionManagerMetrics(instrumentOpts.MetricsScope()),
+	}
+	go mgr.reportMetrics()
+	return mgr
+}
+
+func (mgr *electionManager) Open(shardSetID string) error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	if mgr.state != electionManagerNotOpen {
+		return errElectionManagerExpectNotOpen
+	}
+	mgr.state = electionManagerOpen
+	mgr.electionKey = fmt.Sprintf(mgr.electionKeyFmt, shardSetID)
+	go mgr.startElectionLoop()
+	return nil
+}
+
+func (mgr *electionManager) ElectionState() ElectionState {
+	mgr.RLock()
+	electionState := mgr.electionState
+	mgr.RUnlock()
+	return electionState
+}
+
+func (mgr *electionManager) Resign(ctx context.Context) error {
+	mgr.Lock()
+	if mgr.state != electionManagerOpen {
+		mgr.Unlock()
+		return errElectionManagerExpectOpen
+	}
+	if mgr.electionState == FollowerState {
+		mgr.Unlock()
+		return nil
+	}
+	mgr.state = electionManagerResigning
+	mgr.resignWg = &sync.WaitGroup{}
+	resignWg := mgr.resignWg
+	resignWg.Add(1)
```

Comment:
The increment needs to be done inside the lock, otherwise if the election goroutine does a `resignWg.Wait()` after `mgr.Unlock()` but before `resignWg.Add(1)`, it may proceed prematurely without waiting for the necessary resignation operations to finish.